Recommend DistSender concurrency limit bump (#19096)
cockroachdb/docs: src/current/v24.3/performance-recipes.md

    <td><ul><li>You may be scanning over large numbers of <a href="{% link {{ page.version.version }}/architecture/storage-layer.md %}#mvcc">MVCC versions</a>. This is similar to how a full table scan can be slow.</li></ul></td>
    <td><ul><li><a href="#too-many-mvcc-values">Configure CockroachDB to purge unneeded MVCC values.</a></li></ul></td>
  </tr>
  <tr>
    <td><ul><li>vCPU usage has plateaued (possibly around 70%) on your large cluster.</li></ul></td>

>     <td><ul><li><a href="architecture/distribution-layer.html#distsender">KV layer DistSender</a> batches may be getting throttled; check if the <code>distsender.batches.async.throttled</code> metric is greater than <code>0</code>.</li></ul></td>

Author: unfortunately not yet, it's only in 2 places per [this search result](https://www.cockroachlabs.com/docs/search?query=distsender.batches.async.throttled)

I have filed https://cockroachlabs.atlassian.net/browse/DOC-11752 to get it documented

>     <td><ul><li>Increase the <code>kv.dist_sender.concurrency_limit</code> <a href="cluster-settings.html">cluster setting</a>.</li></ul></td>

Author: we can't (yet) because I just filed an issue to make it public

https://github.com/cockroachdb/cockroach/issues/135615

Reduce the [`gc.ttlseconds`]({% link {{ page.version.version }}/configure-replication-zones.md %}#gc-ttlseconds) zone configuration of the table as much as possible.

### KV DistSender batches being throttled (performance impact to larger clusters)

> If you see `distsender.batches.async.throttled` values that aren't zero (or aren't consistently near zero), experiment with increasing the [KV layer `DistSender`]({% link {{ page.version.version }}/architecture/distribution-layer.md %}#distsender) and `Streamer` concurrency using the `kv.streamer.concurrency_limit` [cluster setting]({% link {{ page.version.version }}/cluster-settings.md %}).  In v24.3, these default values were increased by 6x and 12x, respectively.  For versions prior to v24.3, increasing the value by 6x and 12x would be a good starting point.

Reviewer: ```suggestion
If you see `distsender.batches.async.throttled` values that aren't zero (or aren't consistently near zero), experiment with increasing the [KV layer `DistSender`]({% link {{ page.version.version }}/architecture/distribution-layer.md %}#distsender) and `Streamer` concurrency using the `kv.dist_sender.concurrency_limit` and `kv.streamer.concurrency_limit` [cluster settings]({% link {{ page.version.version }}/cluster-settings.md %}), respectively.  In v24.3, these default values were increased by 6x and 12x, respectively.  For versions prior to v24.3, increasing the value by 6x and 12x would be a good starting point.
```

Author: I have updated to use this text, thanks

> If you see `distsender.batches.async.throttled` values that aren't zero (or aren't consistently near zero), experiment with increasing the [KV layer `DistSender`]({% link {{ page.version.version }}/architecture/distribution-layer.md %}#distsender) and `Streamer` concurrency using the `kv.streamer.concurrency_limit` [cluster setting]({% link {{ page.version.version }}/cluster-settings.md %}).  In v24.3, these default values were increased by 6x and 12x, respectively.  For versions prior to v24.3, increasing the value by 6x and 12x would be a good starting point.

Author: Unfortunately this isn't public yet, I added another note to https://github.com/cockroachdb/cockroach/issues/135615 to get the cluster setting made public

> If you see `distsender.batches.async.throttled` values that aren't zero (or aren't consistently near zero), experiment with increasing the [KV layer `DistSender`]({% link {{ page.version.version }}/architecture/distribution-layer.md %}#distsender) and `Streamer` concurrency using the `kv.dist_sender.concurrency_limit` and `kv.streamer.concurrency_limit` [cluster settings]({% link {{ page.version.version }}/cluster-settings.md %}), respectively.  In v24.3, these default values were increased by 6x and 12x, respectively.  For versions prior to v24.3, increasing the value by 6x and 12x would be a good starting point.

Reviewer: ```suggestion
If you see `distsender.batches.async.throttled` values that aren't zero (or aren't consistently near zero), experiment with increasing the [KV layer `DistSender`]({% link {{ page.version.version }}/architecture/distribution-layer.md %}#distsender) and `Streamer` concurrency using the `kv.dist_sender.concurrency_limit` and `kv.streamer.concurrency_limit` [cluster settings]({% link {{ page.version.version }}/cluster-settings.md %}), respectively.  In v24.3, these default values were increased by 6x and 12x, respectively.  For versions prior to v24.3, increasing the values by 6x and 12x would be a good starting point.
```

Author: ✅

> To validate a successful result, you can increase this value until you see no new throttled requests and no increase in tail latency (e.g. `p99.999`).

Reviewer: It's not clear what "this value" refers to since these seem to be two values. Should it be "increase both values"? Should you always increase them together?

```suggestion
To validate a successful result, you can increase this value until you see no new throttled requests and no increase in tail latency (e.g., `p99.999`).
```

> To validate a successful result, you can increase this value until you see no new throttled requests and no increase in tail latency (e.g., `p99.999`).

Reviewer: I'm still not clear on what "this value" refers to?

Author: updated to "the values of these cluster settings", PTAL

> This does increase the amount of RAM consumption per node to handle the increased concurrency, but it's proportional to the load and an individual flow's memory consumption should not be significant. Bad outcomes include increased tail latency or too much memory consumption with no decrease in the number of throttled requests.

Reviewer: Is it possible to suggest what someone would do in a bad outcome?
